internal/jobcontainers/jobcontainer.go

```diff
@@ -0,0 +1,393 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create process token")
+	}
+
+	cmd := &exec.Cmd{
+		Env:  os.Environ(),
+		Dir:  c.sandboxMount,
+		Path: absPath,
+		Args: args,
+		SysProcAttr: &syscall.SysProcAttr{
+			CreationFlags: windows.CREATE_NEW_PROCESS_GROUP,
+			Token:         syscall.Token(token),
+		},
+	}
+
+	process := newProcess(cmd)
+
+	var (
+		stdin  io.WriteCloser
+		stdout io.ReadCloser
+		stderr io.ReadCloser
+	)
+	// Create process pipes if asked for.
+	if conf.CreateStdInPipe {
+		stdin, err = process.cmd.StdinPipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdin pipe")
+		}
+		process.stdin = stdin
+	}
+
+	if conf.CreateStdOutPipe {
+		stdout, err = process.cmd.StdoutPipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdout pipe")
+		}
+		process.stdout = stdout
+	}
+
+	if conf.CreateStdErrPipe {
+		stderr, err = process.cmd.StderrPipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stderr pipe")
+		}
+		process.stderr = stderr
+	}
+
+	defer func() {
+		if err != nil {
+			process.Close()
+			c.Release(ctx)
+		}
+	}()
+
+	if err = process.Start(); err != nil {
+		return nil, errors.Wrap(err, "failed to start host process")
+	}
+
+	if err = c.job.Assign(uint32(process.Pid())); err != nil {
+		return nil, errors.Wrap(err, "failed to assign process to job object")
+	}
+
+	// Assign the first process made as the init process of the container.
+	c.init.initDoOnce.Do(func() {
+		c.init.proc = process
+		close(c.init.initBlock)
+	})
+
+	// Wait for process exit
+	go c.pollJobMsgs(ctx)
+	go process.waitBackground(ctx)
+	return process, nil
+}
+
+// Release unmounts all of the container layers. Safe to call multiple times, if no storage
+// is mounted this call will just return nil.
+func (c *JobContainer) Release(ctx context.Context) error {
+	if c.sandboxMount != "" {
+		if err := removeSandboxMountPoint(ctx, c.sandboxMount); err != nil {
+			log.G(ctx).WithFields(logrus.Fields{
+				"id":   c.id,
+				"path": c.sandboxMount,
+			}).Warn("failed to remove sandbox volume mount")
+			return errors.Wrap(err, "failed to remove sandbox volume mount path")
+		}
+		if err := layers.UnmountContainerLayers(ctx, c.spec.Windows.LayerFolders, "", nil, layers.UnmountOperationAll); err != nil {
+			log.G(ctx).WithFields(logrus.Fields{
+				"id":    c.id,
+				"paths": c.spec.Windows.LayerFolders,
+			}).Warn("failed to unmount container layers")
+			return errors.Wrap(err, "failed to unmount container layers")
+		}
+		c.sandboxMount = ""
+	}
+	return nil
+}
+
+// Start starts the container. This is a noop as there isn't anything to "start".
+func (c *JobContainer) Start(ctx context.Context) error {
+	return nil
+}
+
+// Close closes any open handles.
+func (c *JobContainer) Close() error {
+	if err := c.job.Close(); err != nil {
+		return err
+	}
+	c.closedWaitOnce.Do(func() {
+		c.waitError = hcs.ErrAlreadyClosed
+		close(c.waitBlock)
+	})
```

Comment:
I don't get this. On the first call to `Close` we set the waitError to `ErrAlreadyClosed` but then closing the waitblock, which would trigger `Wait` below and then would return the waitError, which we set to `ErrAlreadyClosed` already. Am I missing something?

Reply:
This was just mimicking the behavior we have for hcs containers currently. I guess the intention is for waitError to signal that the container was closed before actually running to completion/being explicitly killed. https://github.com/microsoft/hcsshim/blob/ae33b435ce9518e56ed5b01cf9638cd3290bdaac/internal/hcs/errors.go#L258